File: scalafmt-interfaces/src/main/java/org/scalafmt/interfaces/ScalafmtClassLoader.java
package org.scalafmt.interfaces;

/**
 * A classloader that shares only scalafmt-interfaces classes from the parent classloader.
 *
 * This classloader is intended to be used as a parent when class-loading scalafmt-dynamic.
 * By using this classloader as a parent, it's possible to cast runtime instances from
 * the scalafmt-dynamic classloader into `org.scalafmt.interfaces.Scalafmt` from this classlaoder.
 */
public class ScalafmtClassLoader extends ClassLoader {

    private ClassLoader parent;
    public ScalafmtClassLoader(ClassLoader parent) {
        super(null);
        this.parent = parent;
    }

    @Override
    protected Class<?> findClass(String name) throws ClassNotFoundException {
        if (name.startsWith("org.scalafmt.interfaces")) {
            return parent.loadClass(name);
        }  else {
            return super.findClass(name);
        }
    }
}

File: scalafmt-interfaces/native/src/main/scala/org/scalafmt/interfaces/ScalafmtClassLoader.scala
package org.scalafmt.interfaces

/** A classloader that shares only scalafmt-interfaces classes from the parent
  * classloader.
  *
  * This classloader is intended to be used as a parent when class-loading
  * scalafmt-dynamic. By using this classloader as a parent, it's possible to
  * cast runtime instances from the scalafmt-dynamic classloader into
  * `org.scalafmt.interfaces.Scalafmt` from this classloader.
  */
class ScalafmtClassLoader(var parent: ClassLoader) extends ClassLoader(null) {
  @throws[ClassNotFoundException]
  override protected def findClass(name: String) =
    if (name.startsWith("org.scalafmt.interfaces")) parent.loadClass(name)
    else super.findClass(name)
}
